src/main/kotlin/frc/robot/lib/SysIdCommand.kt

```diff
@@ -0,0 +1,190 @@
+     * voltage, and timeout.
+     * @return This [SysIdCommand] instance for chaining.
+     */
+    private fun createRoutine(routineConstants: LoggedSysIdRoutineConstants) =
+        SysIdRoutine(
+            SysIdRoutine.Config(
+                routineConstants.loggedRampRate.get().volts.per(sec),
+                routineConstants.loggedStepVoltage.get().volts,
+                routineConstants.loggedTimeout.get().sec,
+            ) { state: SysIdRoutineLog.State ->
+                Logger.recordOutput("$name/state", state.toString())
+            },
+            SysIdRoutine.Mechanism(
+                subsystem.setVoltageConsumer,
+                null,
+                subsystem
+            )
+        )
+
+    /** Initializes the internal SysId routines from the stored constants. */
+    private fun createRoutinesCommand(): Command =
+        subsystem.runOnce {
+            requireNotNull(forwardRoutineConstants) {
+                "Forward routine constants not initialized!"
+            }
+            requireNotNull(backwardRoutineConstants) {
+                "Backward routine constants not initialized!"
+            }
+
+            routineForwards = createRoutine(forwardRoutineConstants!!)
+            routineBackwards = createRoutine(backwardRoutineConstants!!)
+        }
+
+    /**
+     * Builds the full characterization command sequence:
+     * 1. Initializes routines
+     * 2. Runs forward dynamic
+     * 3. Runs backward dynamic
+     * 4. Runs forward quasistatic
+     * 5. Runs backward quasistatic
+     *
+     * Waits 1 second between each step.
+     *
+     * @return The full [Command] sequence.
+     */
+    fun createSysIDCommand(): Command {
+        return subsystem
+            .defer {
+                Commands.sequence(
+                    createRoutinesCommand(),
+                    routineForwards!!.dynamic(SysIdRoutine.Direction.kForward),
+                    Commands.waitSeconds(TIME_BETWEEN_ROUTINES[sec]),
+                    routineBackwards!!.dynamic(SysIdRoutine.Direction.kReverse),
+                    Commands.waitSeconds(TIME_BETWEEN_ROUTINES[sec]),
+                    routineForwards!!.quasistatic(
+                        SysIdRoutine.Direction.kForward
+                    ),
+                    Commands.waitSeconds(TIME_BETWEEN_ROUTINES[sec]),
+                    routineBackwards!!.quasistatic(
+                        SysIdRoutine.Direction.kReverse
+                    )
+                )
+            }
+            .withName("$name/characterize")
+    }
+
+    /**
+     * Holds the constants used for configuring a [SysIdRoutine], with tunable
+     * logging support.
+     *
+     * @param rampRate The ramp rate for quasistatic tests.
+     * @param stepVoltage The voltage step size for dynamic tests.
+     * @param timeout The timeout after which the routine will stop.
+     */
+    inner class LoggedSysIdRoutineConstants(
```

Comment:
The word "constants" here is a bit misleading -- you use this for tuning, so they are not constant. I'd use WPILib's term -- "config". And, perhaps add "tunable" to the name.